domain/status/service/service.go

```diff
@@ -270,6 +287,58 @@ func (s *LeadershipService) GetApplicationAndUnitStatusesForUnitWithLeader(
 	return applicationDisplayStatus, unitWorkloadStatuses, nil
 }
 
+// SetRelationStatus sets the status of the relation to the status provided.
+// Status may only be set by the application leader.
+// Returns NotFound
+func (s *LeadershipService) SetRelationStatus(
+	ctx context.Context,
+	unitName coreunit.Name,
+	relationUUID corerelation.UUID,
+	info corestatus.StatusInfo,
+) error {
+	// Encode status.
+	relationStatus, err := encodeRelationStatus(info)
+	if err != nil {
+		return errors.Errorf("encoding relation status: %w", err)
+	}
+
+	// Get current status.
+	currentStatus, err := s.st.GetRelationStatus(ctx, relationUUID)
+	if err != nil {
+		return errors.Errorf("getting current relation status: %w", err)
+	}
+
+	// Check we can transition from current status to the new status.
+	err = relationStatusTransitionValid(currentStatus, relationStatus)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// If transitioning from Suspending to Suspended, retain any existing
+	// message so that any previous reason for suspending is retained.
+	if relationStatus.Message == "" &&
+		currentStatus.Status == status.RelationStatusTypeSuspending &&
+		relationStatus.Status == status.RelationStatusTypeSuspended {
+		relationStatus.Message = currentStatus.Message
+	}
+
+	// Get application name for leadership check.
+	_, applicationName, err := s.st.GetApplicationIDAndNameByUnitName(ctx, unitName)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Set the status as the leader.
+	err = s.leaderEnsurer.WithLeader(ctx, applicationName, unitName.String(), func(ctx context.Context) error {
```

Comment:
todo: wrap more of the method functionality inside of the `s.leaderEnsurer.WithLeader` function. No use doing all the other work if the current unit is not the leader.

Reply:
Ah yea, this is a good point, I hadn't considered that we didn't have transactionality here, the 3.6 code didn't have it and I was basing this off moving that into the service layer. 

Even though we could get away with it, I've jsut moved the whole get logic into the state layer. I think it really should be there so that we're not relying on the leadership check for transactionality, that feels a bit iffy to me.